### PR TITLE
P greedy cholesky

### PR DESCRIPTION
**Pull request by FabioTucciarone**
Wie über Element besprochen bis auf zweite f greedy implementierung die ist noch sehr unschön.

### CODE REVIEW

File: src/mapping/FGreedyCutSolver.hpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+  /// n_randon
+  const double _tolF = 1e-10;
+
+  Eigen::Index    _inSize  = 0;
+  Eigen::Index    _outSize = 0;
+  Eigen::MatrixXd _kernelEval;
+
+  const mesh::Mesh::VertexContainer _inputVertices;
+  const mesh::Mesh::VertexContainer _outputVertices;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:70]: Do we really need the hard-copy of the vertex container here? It's potentially the largest data structure preCICE holds. I would suggest (and probably prefer) to put them into the function signature where needed or we share ownership via a `PtrMesh`?

File: src/mapping/FGreedyCutSolver.hpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  RADIAL_BASIS_FUNCTION_T _basisFunction;
+  int _basisSize;
+  std::array<bool, 3> _activeAxis;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:73]: I could imagine that the active axis are not required at all here, more a historic relic.

File: src/mapping/FGreedyCholeskySolver.hpp
Comment (davidscn) [inline src/mapping/FGreedyCholeskySolver.hpp:1]: Is this file required? It was really only a draft without the intention to ever merge it somewhere. How is the Cholesky solver different from the Cut solver? The Cut should use a Cholesky decomposition as well, no?

File: src/mapping/FGreedyCutSolver.hpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+template <typename IndexContainer>
+FGreedySolver<RADIAL_BASIS_FUNCTION_T>::FGreedySolver(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::Mesh &inputMesh, const IndexContainer &inputIDs,
+                                                      const mesh::Mesh &outputMesh, const IndexContainer &outputIDs, std::vector<bool> deadAxis, Polynomial polynomial)
+  : _inSize(inputMesh.vertices().size()), 
+    _outSize(outputMesh.vertices().size()), 
+    _inputVertices(inputMesh.vertices()),
+    _outputVertices(outputMesh.vertices()),
+    _basisFunction(basisFunction)
+{
+  precice::profiling::Event initEvent("FGreedy.initialize", profiling::Synchronize);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:146]:
```suggestion
  precice::profiling::Event initEvent("FGreedy.initialize");
```
Note that the current mapping class (`RadialBasisFctMapping`), which is instantiated with this solver uses a gather-scatter approach to collect distributed data and the solver is only executed in serial. The `profiling::Synchronize` is an [MPI barrier](https://docs.open-mpi.org/en/v5.0.x/man-openmpi/man3/MPI_Barrier.3.html). Thus, the sync would lead to a deadlock here in case of parallel execution with synchronization.

File: src/mapping/FGreedyCutSolver.hpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+  PRECICE_ASSERT(polynomial == Polynomial::OFF, "Poly off");
+  PRECICE_ASSERT(RADIAL_BASIS_FUNCTION_T::isStrictlyPositiveDefinite());
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:149]: Why does this code not work for c.p.d. kernels?

File: src/mapping/FGreedyCutSolver.hpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+
+// ---- Evaluation ---- //
+
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd FGreedySolver<RADIAL_BASIS_FUNCTION_T>::solveConservative(const Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  // Not implemented
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:167]: I think this case would be possible though, I have to think about it.

File: src/mapping/FGreedyCutSolver.hpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                                                      const mesh::Mesh &outputMesh, const IndexContainer &outputIDs, std::vector<bool> deadAxis, Polynomial polynomial)
+  : _inSize(inputMesh.vertices().size()), 
+    _outSize(outputMesh.vertices().size()), 
+    _inputVertices(inputMesh.vertices()),
+    _outputVertices(outputMesh.vertices()),
+    _basisFunction(basisFunction)
+{
+  precice::profiling::Event initEvent("FGreedy.initialize", profiling::Synchronize);
+
+  PRECICE_ASSERT(polynomial == Polynomial::OFF, "Poly off");
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:148]: I think we could still add the polynomial interpolation in our case, and 'only' replace the kernel interpolation with the greedy approach, but it would require some experiments to see if this pays off. In general, computing and evaluating the polynomial part is computationally cheap.

File: src/mapping/FGreedyCutSolver.hpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+  PRECICE_ASSERT(RADIAL_BASIS_FUNCTION_T::isStrictlyPositiveDefinite());
+  PRECICE_ASSERT(_kernelEval.size() == 0); 
+
+  _basisSize = std::min(static_cast<int>(_inSize), _maxIter);
+
+  _activeAxis = std::array<bool, 3>({{false, false, false}});
+  std::transform(deadAxis.begin(), deadAxis.end(), _activeAxis.begin(), [](const auto ax) { return !ax; });
+
+  initEvent.stop();
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:157]:
```suggestion
```
Events stop automatically, when the destructor is being called, i.e., we leave the the current scope.

File: src/mapping/FGreedyCutSolver.hpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+namespace mapping {
+
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+class FGreedySolver {
+public:
+  using DecompositionType = std::conditional_t<RADIAL_BASIS_FUNCTION_T::isStrictlyPositiveDefinite(), Eigen::LLT<Eigen::MatrixXd>, Eigen::ColPivHouseholderQR<Eigen::MatrixXd>>;
+  using BASIS_FUNCTION_T  = RADIAL_BASIS_FUNCTION_T;
+  /// Default constructor
+  FGreedySolver() = default;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:26]:
```suggestion
  FGreedySolver() = default;
```
I think that's not required anymore (it was in the past).

File: src/mapping/FGreedyCutSolver.hpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+#include "mapping/RadialBasisFctSolver.hpp"
+#include "mesh/Mesh.hpp"
+#include "precice/impl/Types.hpp"
+#include "profiling/Event.hpp"
+#include <fenv.h>
+
+namespace precice {
+namespace mapping {
+
+
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:19]: Could you add some documentation here on what this solver is about. It should be understandable for the others working with the code, but not necessarily with data mapping.

File: src/mapping/FGreedyCutSolver.hpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  return Eigen::VectorXd();
+}
+
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd FGreedySolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  precice::profiling::Event solveEvent("FGreedy.solveConsistent", profiling::Synchronize);
+
+  Eigen::VectorXd residual = inputData;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:178]: The inputData is non-const and I think you can happily reshape it as needed, so there is no additional copy needed.

File: src/mapping/FGreedyCutSolver.hpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+  Eigen::MatrixXd kernelMatrix = Eigen::MatrixXd::Zero(_inSize, _basisSize);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:179]: Ok, according to your code, the `_basisSize` is an upper limit of the matrix (and also the following) vector sizes. I assume that this is now a question on how to use the code from the user perspective:

I would assume that the maximum number of iterations is implicitly `inSize`. How large can the cases be we can compute with this solver (of course we run with the RBF solver into a memory limit as well). If I now want to test a larger case, let's say $10^5$ in vertices, than the code will die here due to all the memory allocations, which might, however, not be necessary at all if my basis size is significantly smaller, right?

File: src/mapping/FGreedyCutSolver.hpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  Eigen::MatrixXd cut = Eigen::MatrixXd::Zero(_basisSize, _basisSize);
+  std::vector<int> greedyIDs;
+  greedyIDs.reserve(_basisSize);
+  Eigen::VectorXd v = Eigen::VectorXd::Ones(_basisSize);
+  Eigen::VectorXd v2 = Eigen::VectorXd::Ones(_basisSize);
+  Eigen::VectorXd c = Eigen::VectorXd::Zero(_inSize);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:185]: I assume you took the variable names from the python code, but maybe you can find more expressive names for these variables?

File: src/mapping/FGreedyCutSolver.hpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+  // Iterative selection of new points
+  for (int n = 0; n < _basisSize; ++n) {
+
+    auto [i, fMax] = select(residual);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:190]: From what I understood from the python code, a fundamental design decision for the code was essentially that only the `selectionRule` is different between P and F greedy algorithm? How does this apply for us? Could we move this code part into a separate helper and use it for both solvers or are there major performance issues with the P greedy then?

File: src/mapping/FGreedyCutSolver.hpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+  }
+  const double a = (cut.block(n-1,0,1,n) * inputData(greedyIDs))(0,0);
+  c.head(n) += a * cut.block(n-1,0,1,n).transpose();
+  return (inputData - kernelMatrix.block(0,0,_inSize, n) * c.head(n)).cwiseAbs();
+}
+
+
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+void FGreedySolver<RADIAL_BASIS_FUNCTION_T>::updateKernelVector(const std::vector<int> greedyIDs, Eigen::VectorXd &kernelVector, const mesh::Vertex &x) const
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:123]: Could you swap the argument order here to have the 'return by reference' as the last entry?

File: src/mapping/FGreedyCutSolver.hpp [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd FGreedySolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  precice::profiling::Event solveEvent("FGreedy.solveConsistent", profiling::Synchronize);
+
+  Eigen::VectorXd residual = inputData;
+  Eigen::MatrixXd kernelMatrix = Eigen::MatrixXd::Zero(_inSize, _basisSize);
+  Eigen::MatrixXd cut = Eigen::MatrixXd::Zero(_basisSize, _basisSize);
+  std::vector<int> greedyIDs;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:181]: We could also consider to make the `greedyIDs` an `Eigen::VectorXi` to have it more consistent here.

File: src/mapping/FGreedyCutSolver.hpp [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+    auto [i, fMax] = select(residual);
+    auto x = _inputVertices.at(i); 
+    greedyIDs.push_back(i);
+
+    if (fMax < _tolF) break;
+
+    updateKernelVector(greedyIDs, v, x);
+    v2.head(n) = cut.block(0, 0, n, n).triangularView<Eigen::Lower>() * v.head(n);
+    double p = std::sqrt(v2(n));  
+    v.head(n) = v2.head(n) / p;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:199]:
```suggestion
    double p_inv = 1. / std::sqrt(v2(n));  
    v.head(n) = v2.head(n) * p_inv;
```
Of course measuring would make sense, but this seems to be more intuitive.

File: src/mapping/FGreedyCutSolver.hpp [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+    cut.block(n, 0, 1, n).noalias() = -v.block(0, 0, n, 1).transpose() * cut.block(0, 0, n, n).triangularView<Eigen::Lower>();
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:201]:
```suggestion
    cut.block(n, 0, 1, n).noalias() = -v.head(n).transpose() * cut.block(0, 0, n, n).triangularView<Eigen::Lower>();
```
No? you use this already above? The first cut is essentially one column, cropped to the specific limits of n, right?

File: src/mapping/FGreedyCutSolver.hpp [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+
+    auto [i, fMax] = select(residual);
+    auto x = _inputVertices.at(i); 
+    greedyIDs.push_back(i);
+
+    if (fMax < _tolF) break;
+
+    updateKernelVector(greedyIDs, v, x);
+    v2.head(n) = cut.block(0, 0, n, n).triangularView<Eigen::Lower>() * v.head(n);
+    double p = std::sqrt(v2(n));
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:198]:
```suggestion
    PRECICE_ASSERT(v2(n)>0);
    double p = std::sqrt(v2(n));  
```

File: src/mapping/FGreedyCutSolver.hpp [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+    if (fMax < _tolF) break;
+
+    updateKernelVector(greedyIDs, v, x);
+    v2.head(n) = cut.block(0, 0, n, n).triangularView<Eigen::Lower>() * v.head(n);
+    double p = std::sqrt(v2(n));  
+    v.head(n) = v2.head(n) / p;
+
+    cut.block(n, 0, 1, n).noalias() = -v.block(0, 0, n, 1).transpose() * cut.block(0, 0, n, n).triangularView<Eigen::Lower>();
+    cut(n, n) = 1;
+    cut.block(n, 0, 1, n + 1) /= v(n);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:203]: Maybe (again) pre-compute the inverse of v(n)?

File: src/mapping/FGreedyCutSolver.hpp [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd FGreedySolver<RADIAL_BASIS_FUNCTION_T>::recalculateResidual(const std::vector<int> &greedyIDs, const Eigen::MatrixXd &cut, Eigen::MatrixXd &kernelMatrix, const Eigen::VectorXd &inputData, Eigen::VectorXd &c) const {
+  int n = greedyIDs.size();
+  const auto &v = _inputVertices.at(greedyIDs.at(n-1)).rawCoords();
+  for (size_t i = 0; i < _inSize; i++) 
+  {
+    const auto &u = _inputVertices.at(i).rawCoords();
+    const double squaredDifference = computeSquaredDifference(u, v, _activeAxis);
+    kernelMatrix(i, n-1) = _basisFunction.evaluate(std::sqrt(squaredDifference));
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:113]: Looks somewhat similar to the `updateKernelVector`. The information is not computed elsewhere, right?

File: src/mapping/FGreedyCutSolver.hpp [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+  }
+  const double a = (cut.block(n-1,0,1,n) * inputData(greedyIDs))(0,0);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:115]: Is `(0,0)` just accessing the very first element or referring to the starting row and col?

File: src/mapping/FGreedyCutSolver.hpp [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+    updateKernelVector(greedyIDs, v, x);
+    v2.head(n) = cut.block(0, 0, n, n).triangularView<Eigen::Lower>() * v.head(n);
+    double p = std::sqrt(v2(n));  
+    v.head(n) = v2.head(n) / p;
+
+    cut.block(n, 0, 1, n).noalias() = -v.block(0, 0, n, 1).transpose() * cut.block(0, 0, n, n).triangularView<Eigen::Lower>();
+    cut(n, n) = 1;
+    cut.block(n, 0, 1, n + 1) /= v(n);
+
+    residual = recalculateResidual(greedyIDs, cut, kernelMatrix, inputData, c);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:205]: Would it make sense to put the residual into the function to avoid reallocations here?

File: src/mapping/FGreedyCutSolver.hpp [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+
+    std::cout << "iteration = " << n << "\r";
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:207]:
```suggestion
   PRECICE_DEBUG( "Iteration number: {}", n);
```
Will be later of relevance for debugging. Maybe logging the fMax as well might be of interest.

File: src/mapping/FGreedyCutSolver.hpp [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+  }
+
+  Eigen::MatrixXd kernelEval = buildEvaluationMatrix(greedyIDs);
+  
+  Eigen::VectorXd y = inputData(greedyIDs);
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:212]:
```suggestion
```
?

File: src/mapping/FGreedyCutSolver.hpp [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+  Eigen::VectorXd prediction = kernelEval.transpose() * c;
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:213]: Instead of really allocating and building the whole kernelEval matrix which we throw away afterwards, one could consider to merge the multiplication into the `buildEvaluationMatrix` function and then compute the predictions component-wise withing the functions. This could make the code improve data locality and reduce the memory footprint, but of course, requires measuring if this is critical here.

File: src/mapping/FGreedyCutSolver.hpp [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+
+  solveEvent.stop();
```
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:215]:
```suggestion
```

File: src/mapping/config/MappingConfiguration.cpp [hunk 27]
```diff
@@ -62,7 +62,7 @@ void addAttributes(TagStorage &storage, const std::vector<variant_t> &attributes
 // Enum required for the RBF instantiations
 enum struct RBFBackend {
   Eigen,
-  EigenGreedy,  //TODO: Das ist unschön. Aber intern die Methoder durch das Backend zu bestimmen auch!
+  EigenGreedy,  //TODO: Das ist unschön.
```
Comment (davidscn) [inline src/mapping/config/MappingConfiguration.cpp:65]: Ideally we would have here something like `Greedy` and the exact algorithm (P or f) remains configurable.

File: src/mapping/PGreedyCutSolver.hpp
Comment (davidscn) [inline src/mapping/PGreedyCutSolver.hpp:1]: This is then my drafted solver? Not sure why this file is not new? Should I take a look here as well?

File: src/mapping/FGreedyCutSolver.hpp
Comment (davidscn) [inline src/mapping/FGreedyCutSolver.hpp:1]: Is it possible (for the F greedy) to have the center selection only once every couple of iterations? The code currently solves all centers during `solveConsistent` and recomputes them completely..

**Review by davidscn — COMMENTED** [review]
Looks great so far :) Some comments below, do you also have the results for the turbine you performed so far?

You can use the script provided in `./tools/formatting/format-all` to indent the code in a preCICE-conforming way and make the code more reader-friendly. The script uses an ancient `clang-format`  version 6.

File: src/mapping/FGreedyCholeskySolver.hpp
Comment (davidscn) [reply src/mapping/FGreedyCholeskySolver.hpp:1]: In my last review this was shown as a renamed file..

File: src/mapping/config/MappingConfiguration.cpp [hunk 28]
```diff
@@ -247,6 +256,7 @@ MappingConfiguration::MappingConfiguration(
   // Add the relevant attributes to the relevant tags
   addAttributes(projectionTags, {attrFromMesh, attrToMesh, attrDirection, attrConstraint});
   addAttributes(rbfDirectTags, {attrFromMesh, attrToMesh, attrDirection, attrConstraint, attrPolynomial, attrXDead, attrYDead, attrZDead});
+  addAttributes(rbfGreedyTags, {attrFromMesh, attrToMesh, attrDirection, attrConstraint, attrPolynomial, attrXDead, attrYDead, attrZDead}); // TODO: Hier Tag für Greedy Typen registrieren. //TODO: prüfe ob attrPolynomial, ... notwendig?
```
Comment (davidscn) [inline src/mapping/config/MappingConfiguration.cpp:259]: See above, we could definitely try to combine greedy and polynomials

File: tests/serial/mapping-greedy-kernel/GreedyDirection1.xml [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+
+  <participant name="SolverOne">
+    <provide-mesh name="MeshOne" />
+    <write-data name="DataOne" mesh="MeshOne" />
+  </participant>
+
+  <participant name="SolverTwo">
+    <receive-mesh name="MeshOne" from="SolverOne" />
+    <provide-mesh name="MeshTwo" />
+    <mapping:rbf-greedy
```
Comment (davidscn) [inline tests/serial/mapping-greedy-kernel/GreedyDirection1.xml:21]: Maybe an additional attribute for the tolerance (fMax) ? Can we give a default for that?

File: tests/serial/mapping-greedy-kernel/helpers.cpp [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+
+void testGreedyMappingDirection1(const std::string configFile, const TestContext &context)
+{
+  using Eigen::Vector3d;
+
+  std::vector<double> values;
+  for (unsigned int i = 0; i < 12; ++i)
+    values.emplace_back(std::pow(i + 1, 2));
+
+  double expectedValues[3] = {1.0000000000000002, 2.1879131472090689, 22.859664317930537};
```
Comment (davidscn) [inline tests/serial/mapping-greedy-kernel/helpers.cpp:49]:
```suggestion
  std::array<double, 3> expectedValues({{1.0000000000000002, 2.1879131472090689, 22.859664317930537}});
```

File: tests/serial/mapping-greedy-kernel/helpers.cpp [hunk 31]
```diff
@@ -0,0 +1,182 @@
...
+    // Initialize, thus receive the data and map.
+    interface.initialize();
+    double maxDt = interface.getMaxTimeStepSize();
+    BOOST_TEST(interface.isCouplingOngoing(), "Receiving participant should have to advance once!");
+
+    // Read the mapped data from the mesh.
+    auto dataAID = "DataOne";
+    BOOST_TEST(!interface.requiresGradientDataFor(meshTwoID, dataAID));
+
+    double values[3];
```
Comment (davidscn) [inline tests/serial/mapping-greedy-kernel/helpers.cpp:92]:
```suggestion
    std::array<double,3> values;
```

File: tests/serial/mapping-greedy-kernel/helpers.cpp [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+
+  auto meshOneID = "MeshOne";
+  auto meshTwoID = "MeshTwo";
+  auto dataAID = "DataOne";
+
+  std::vector<double> values;
+  for (unsigned int i = 0; i < 3; ++i)
+    values.emplace_back(std::pow(i + 1, 2));
+
+  double expectedValues[12] = {
```
Comment (davidscn) [inline tests/serial/mapping-greedy-kernel/helpers.cpp:120]: I assume you put these values as they were generated by the algorithm, right? It would be good to also have some more simple test cases, for which you know the results analytically, e.g., if an input vertex is also in the output set, then input should be equal to output, or if we have a linear function between two vertices and evaluate in the middle, there should be a reasonable guess on the result. We could also consider to perform regression testing against the `rbf-direct` solver. Maybe you can simply re-use some of the existing solver tests we have https://github.com/precice/precice/blob/develop/src/mapping/tests/GinkgoRadialBasisFctSolverTest.cpp

This makes debugging more easy in case test cases fail.

File: src/mapping/FGreedyCholeskySolver.hpp
Comment (FabioTucciarone) [reply src/mapping/FGreedyCholeskySolver.hpp:1]: The Cholesky decomposition of the kernel matrix A on the input mesh is given by the Newton-Basis Matrix V: A = VV^T. In the Code _cut is the inverse of V.

The Cholesky Solver updates the columns of the complete decomposition matrix and adds the newly completed rows to the result. It does not compute the necessary information for _cut.

The Cut Solver computes only a part of the full decomposition matrix and uses it to extend the inverse. But it does not accumulate the necessary information to update the greedy criteria and the residual. These need to be partially recomputed.
A Cholesky decomposition could also be calculated here but would be an additional (unnecessary) result as the _cut matrix is necessary for each iteration.

File: src/mapping/FGreedyCutSolver.hpp
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:1]: This is possible for f-greedy as well as P-greedy (Cut Solver). One would need to check how f and P behave in the vicinity of a newly added point and if this is a sensible strategy.

File: tests/serial/mapping-greedy-kernel/GreedyDirection1.xml (hunk 29 above)
Comment (FabioTucciarone) [reply tests/serial/mapping-greedy-kernel/GreedyDirection1.xml:21]: Added parameters for a maximum number of iterations and the tolerance.

File: tests/serial/mapping-greedy-kernel/helpers.cpp (hunk 32 above)
Comment (FabioTucciarone) [reply tests/serial/mapping-greedy-kernel/helpers.cpp:120]: The values were generated by rbf-global-direct. The greedy algorithms should interpolate the function exactly the same way as rbf-global-direct when including every input vertex.

File: src/mapping/FGreedyCutSolver.hpp (hunk 24 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:212]: y was used for testing and forgotten.

File: src/mapping/FGreedyCutSolver.hpp (hunk 21 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:115]: Converts the 1x1 Matrix into a double. Replaced with .value().

File: src/mapping/FGreedyCutSolver.hpp (hunk 1 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:70]: RadialBasisFctMapping passes the meshes to the function via reference. The objects are destroyed by the time the solve methods are called (directly after the initialization of the solver, which happens at the end of the function). Replaced copy with move.

File: src/mapping/FGreedyCutSolver.hpp (hunk 2 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:73]: Should I remove _activeAxis then?

File: src/mapping/FGreedyCutSolver.hpp (hunk 5 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:167]: Yes, I don't see any problems with it. It just didn't have priority.

File: src/mapping/FGreedyCutSolver.hpp (hunk 10 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:178]: The unmodified inputData is still needed for the update of the residual.

File: src/mapping/FGreedyCutSolver.hpp (hunk 12 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:185]: Everything is sensibly renamed now.

File: src/mapping/FGreedyCutSolver.hpp (hunk 13 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:190]: The select method can probably be moved, it only searches for a maximum and calculates the absolute values for FGreedyCholeskySolver.

File: src/mapping/FGreedyCutSolver.hpp (hunk 15 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:181]: I'd rather not use Eigen::VectorXi because the greedyIDs are conceptually different from a mathematical vector. In pseudocode notation, you would consider them to be a set.

File: src/mapping/FGreedyCutSolver.hpp (hunk 20 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:113]: The code is almost the same, but updateKernelVector iterates through different indices.

File: src/mapping/config/MappingConfiguration.cpp (hunk 28 above)
Comment (FabioTucciarone) [reply src/mapping/config/MappingConfiguration.cpp:259]: Added the parameter back.

File: src/mapping/FGreedyCutSolver.hpp (hunk 17 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:201]: Using .head(n) produces "error: static assertion failed: YOU_TRIED_CALLING_A_VECTOR_METHOD_ON_A_MATRIX". I assume because, even though it is transposed, the vector is on the left.

Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:201]: cut is always a square matrix and gets updated the following way:
grafik

File: src/mapping/FGreedyCutSolver.hpp (hunk 4 above)
Comment (FabioTucciarone) [reply src/mapping/FGreedyCutSolver.hpp:149]: It works for all positive definite kernels as defined here in definition 1.:
https://www.researchgate.net/publication/245120269_An_Improved_Vectorial_Kernel_Orthogonal_Greedy_Algorithm

File: src/mapping/FGreedyCutSolver.hpp (hunk 20 above)
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:113]: Then maybe parametrize the IDs or just make it accept iterators or something?

File: src/mapping/FGreedyCutSolver.hpp
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:1]: We need a bit more care with the word 'iteration'. In this context, iteration referred to coupling iterations of my simulation, i.e., I have one set of centers selected and then use the same set of greedy centers in the `solveConsistent` step. I think you understood it as the 'iteration' used to select the centers as such.

File: src/mapping/FGreedyCutSolver.hpp (hunk 1 above)
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:70]: Hm I see. Then move vs copy makes only very little difference. 

The main reason this happens is due to the parallelization: `RadialBasisFctMapping` collects the distributed mesh data structure on the primary rank in a temporary mesh and then solved it (always) in serial on the primary rank. This makes it in practice of course even more painful to copy the mesh over. If we restrict our attention to serial execution, we could just store a pointer to the (persistent) mesh data structure preCICE operates on

https://github.com/precice/precice/blob/361fc2eae4eade0b8e77fd6903c28f38fb16ad81/src/mapping/RadialBasisFctMapping.hpp#L152-L155

Then we keep it as is for now. This is essentially a question about parallelization.

File: src/mapping/FGreedyCutSolver.hpp (hunk 2 above)
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:73]: Yes, there should be a default value where you might need it and we could add it again if it turns about to be required.

File: src/mapping/FGreedyCutSolver.hpp (hunk 15 above)
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:181]: I see your point. Depending on which kind of vector manipulation you need to perform with the IDs, the eigen version might be more feature rich though, potentially in terms of slicing or array operations on other matrices.

File: src/mapping/FGreedyCutSolver.hpp (hunk 17 above)
Comment (davidscn) [reply src/mapping/FGreedyCutSolver.hpp:201]: Not sure if this makes any difference, just note that (in such a case) you can also give indices at compile time

https://libeigen.gitlab.io/docs/group__Core__Module.html#gac01f234bce100e39e6928fdc470e5194
